crates/ide_assists/src/handlers/lambdify_for_each.rs

```diff
@@ -0,0 +1,388 @@
+
+            builder.replace(range, buf)
+        },
+    )
+}
+
+// Assist: convert_sum_call
+//
+// Converts a sum into a sum().
+// ```
+// # //- minicore: iterators
+// # use core::iter;
+// fn main() {
+//     let it = core::iter::repeat(92);
+//     let mut val: usize = 0;
+//     it.for_each$0(|x| val += x);
+// }
+// ```
+// ->
+// ```
+// # use core::iter;
+// fn main() {
+//     let it = core::iter::repeat(92);
+//     let mut val: usize = 0;
+//     val += it.sum::<usize>();
+// }
+// ```
+pub(crate) fn convert_sum_call(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let method = ctx.find_node_at_offset::<ast::MethodCallExpr>()?;
+
+    let closure = match method.arg_list()?.args().next()? {
+        ast::Expr::ClosureExpr(expr) => expr,
+        _ => return None,
+    };
+
+    let (method, receiver) = validate_method_call_expr(ctx, method)?;
+
+    let param_list = closure.param_list()?;
+    let param = param_list.params().next()?.pat()?;
+    let body = closure.body()?;
+
+    let range = method.syntax().text_range();
+    let module = ctx.sema.scope(param.syntax()).module()?;
+
+    let binexpr = match body.clone() {
+        Expr::BinExpr(expr) => expr,
+        Expr::BlockExpr(block) => {
+            let mut stmts = block.statements();
+            let fst_stmt = stmts.next()?;
+            continue_iff(stmts.next().is_none())?; // Only one statement
+            // First statement is an expression...
+            let expr_stmt = match fst_stmt {
+                ast::Stmt::ExprStmt(expr_stmt) => expr_stmt,
+                _ => return None,
+            };
+
+            // ...and even a binary expr...
+            let expr = expr_stmt.expr()?;
+            let my_bin_expr = match expr {
+                ast::Expr::BinExpr(my_bin_expr) => my_bin_expr,
+                _ => return None,
+            };
+            my_bin_expr
+        },
+        _ => return None,
+    };
+    let op = match binexpr.op_kind()? {
+        BinaryOp::Assignment { op } => op?,
+        _ => return None,
+    };
+    match op {
+        ArithOp::Add => (),
+        _ => return None,
+    }
+    continue_iff(format!("{}", binexpr.rhs()?) == format!("{}", param))?;
+    let sum = binexpr.lhs()?;
+
+    let ty = ctx.sema.type_of_pat(&param)?.adjusted();
+
+    // Fully unresolved or unnameable types can't be annotated
+    if (ty.contains_unknown() && ty.type_arguments().count() == 0) || ty.is_closure() {
+        return None;
+    }
+
+    let inferred_type = ty.display_source_code(ctx.db(), module.into()).ok()?;
+
+    acc.add(
+        AssistId("convert_sum_call", AssistKind::RefactorRewrite),
+        "Replace this sum in disguise with a `sum()` call",
+        range,
+        |builder| {
+            let mut buf = String::new();
+            format_to!(buf, "{} += {}.sum::<{}>()", sum, receiver, inferred_type);
+            builder.replace(range, buf)
+        },
+    )
+}
+
+// Assist: convert_all_call
```

Comment:
similar as to the `sum` one, it is unlikely that you'll have a for each exactly as the one here required